rfcs/text/0014_notification_service.md

```diff
@@ -0,0 +1,373 @@
+  Having more use-cases, we can consider extending notification creation API to specify Kibana-specific permissions for the recipients
+  (send a notification to all the users with access to a feature X or space Y).
+  Although, it will require declaring Notification plugin dependency on other Kibana plugins, which can turn into circular dependency problems later.
+- Kibana calculates whether a user should receive a notification in runtime before showing one to the customer.
+  The only available option for the current Kibana Security model. Considering the future integration with UNS,
+  Kibana has to provide an HTTP API to check users' permissions are sufficient to see a notification.
+  Unified Notification center performs a call to HTTP API to filter out notifications before rendering them in Cloud UI. 
+  This option adds a significant runtime performance penalty. Moreover, it makes the rendering process dependent on Kibana instance availability. If one of them is temporarily unavailable, the UNS can't decide if a notification must be shown or not.
+
+### Notification System
+Abstracts the way notifications are created, stored, configured, and retrieved.
+It allows Kibana to adopt different strategies for on-premise and Cloud versions:
+- *LocalRepository* - a repository for on-premise version, it stores data within the current deployment. 
+- *RemoteRepository* - a repository for the Cloud version of Kibana. It stores data within the Unified Notification System deployed in the Cloud.
+In this case, a user can see in Kibana UI the external notifications created outside of Kibana.
+
+*Repository* interface:
+- `create(NotificationEvent)`- creates a notification in the Notification system. 
+- `get({ search_after?: number; size?: number; type_id?: string[] })` - retrieves last notifications for the current user.
+Might be paginated using `search_after` filed.
+- `markAsRead({ notification_id: string[] })` - mark notifications with given ids as read.
+- `markAsUnread({ notification_id: string[] })` - mark notifications with given ids as unread.
+- `markAllAsRead()` - mark all the notifications as read.
+- `markAsPin({ notification_id: string[] })` - mark notifications with given ids as pinned.
+- `markAsUnpin({ notification_id: string[] })` - mark notifications with given ids as unpinned.
+- `configure()` - updates user-specific notification settings.
+
+#### LocalRepository
+![image](../images/0014/local_repository.png)
+
+##### Notification model
+Notification model is responsible for:
+- Applying policies to an incoming notification. NS must check whether a recipient unsubscribed from a source before processing to delivery.
+- Enhancing an incoming notification with necessary data: 
+    - `expired_at` if not specified. Kibana allows administrators to set how long notifications will be stored. Falls back to 30 - 90 days be default. 
+- Writing new incoming messages to the Notification storage.
```

Comment:
no, we shouldn't store such notifications at all.

Reply:
I wonder if that might have any side-effects. I'm thinking about the following scenario:

1. A plugin creates the notification (i.e.: "Upgrade now to 7.11.0! 🎉"). 
2. The plugin wants to get the notification to edit it (7.11.1 came out before it was even read), or delete it because it's no longer valid (the user upgraded to 7.11.0 so there's no point in making noise about it), or to report telemetry about their notifications being read/seen/discarded.
3. We don't store them because the user unsubscribed from that "releases" channel.
4. The plugin gets 404 to all of the requests in [2].

Is that a valid scenario? WDYT?

Reply:
> The plugin wants to get the notification to edit it (7.11.1 came out before it was even read), or delete it because it's no longer valid (the user upgraded to 7.11.0 so there's no point in making noise about it)

I’m not sure that we should support `delete()`API. It might lead to a confusing UX when a notification suddenly disappears from the notification list. What if a user wants to update to `7.11.0`, but not `7.12.0`?  Even in this case, a notification source is able to delete the old notification via setting `expire_at` to the `7.12.0` release date.

In theory, we can add API to delete *all* notifications for a user `deleteAllForUser` for a use-case when a user profile is deleted.
Even though I assumed it’s better to keep the API surface small and allow the Notification system to delete notifications once they are expired.
Do you see a reason why it wouldn’t work?

Reply:
> I’m not sure that we should support `delete()`API. It might lead to a confusing UX when a notification suddenly disappears from the notification list. What if a user wants to update to `7.11.0`, but not `7.12.0`? Even in this case, a notification source is able to delete the old notification via setting `expire_at` to the `7.12.0` release date.

What you're saying makes sense to me. Although we've had situations when we've pushed forward/back a release. I'm even thinking about `7.11.1` coming sooner than expected for a hotfix. We can't anticipate that when setting the `expire_at` on the creation of the notification. But, as you mentioned: the user might still want to react to the notification in `7.11.0` (or even read the changelog that might be included in it vs. only reading the one for 7.11.1).

I can think of another use case: in Cases, we might want to _edit_ the notification regarding one case. This way it always shows the latest status (similar to what GH does). Otherwise, we might create one notification per comment/event/assigned + the closed status. What do you think? Are we OK with listing them all in chronological order, mixing types of notifications and cases? Should we consider a _Grouping_ logic then so all the notifications related to one case are grouped together? Maybe this is also a question to PMs rather than us deciding about it 😇 

> In theory, we can add API to delete _all_ notifications for a user `deleteAllForUser` for a use-case when a user profile is deleted.

This is an interesting point! Maybe something to be provided by the User's storage once it's implemented? I guess this same problem will exist for any other future scenarios that will use the user-specific storage/settings? What do you think?

>  Even though I assumed it’s better to keep the API surface small and allow the Notification system to delete notifications once they are expired.

I agree with the _Keep it simple_ approach. No need to implement it in the first approach. Just thinking ahead, in case it's something we want to support in the future because it might require different approaches if/how we want to support them.

Reply:
> Are we OK with listing them all in chronological order, mixing types of notifications and cases? Should we consider a Grouping logic then so all the notifications related to one case are grouped together?

That's a valid case. We do have `grouping` by `group_id` for this. It's just not covered well in the current RFC. 
I will add an example. In nutshell: a user sees in Flyout only the last notification within a group and can expand the whole group by clicking on `+ 3 messages` See https://github.com/elastic/kibana/pull/90297#discussion_r574698044

Reply:
All good then! Thank you! I'll resolve this :)